### PR TITLE
[BEAM-6058] Adding option for flink configuration directory and setting config in exectution environment

### PR DESCRIPTION
**Pull request by angoenka**
This is needed as when submitting to remote execution environment, client config from the request is used.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by angoenka** [comment]
Run Python Flink ValidatesRunner

**Comment by angoenka** [comment]
Run Flink ValidatesRunner

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 1]
```diff
@@ -204,4 +205,10 @@
   ExecutionMode getExecutionModeForBatch();
 
   void setExecutionModeForBatch(ExecutionMode executionMode);
+
+  @Description("Directory containing Flink YAML configuration files.")
+  @Nullable
+  String getFlinkConfigDir();
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Looks good. Could you test the default in `PipelineOptionstest#testDefaults`?

**Comment by mxm** [comment]
I think this is required to load a custom config for portability. If you submit the "legacy" way, you would have a Flink installation with a custom config which would be automatically discovered.

**Comment by tweise** [comment]
> I think this is required to load a custom config for portability. If you submit the "legacy" way, you would have a Flink installation with a custom config which would be automatically discovered.

Same applies in either case (which is why we need documentation). In portability case the job server replaces the legacy runner and it would discover that configuration (that's how we run the job server in Lyft).

**Comment by mxm** [comment]
I agree that we should document the reason for changes in JIRA.

How do you let the JobServer discover the config? Via the Java property?

**Comment by tweise** [comment]
The machine has Flink installed and `flink-conf.yaml` is same as on every cluster node. No extra properties are required for the job server (other than `--flink-master-url` or eq pipeline option).

**Comment by angoenka** [comment]
Properties like taskmanager.runtime.large-record-handler are read from the submitted JobGraph and are not read by cluster when using RemoteEnvironment. Hence we need to read and set the properties in the job graph. Many of these properties are not even exposed via ExecutionConfig so we need to read these properties directly. 
Reading from a config dir seems be the right way as we are already doing it for parallelism.

**Comment by angoenka** [comment]
Also updated the jira with more details.

**Comment by angoenka** [comment]
@tweise Reminder for review :)

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 2]
```diff
@@ -204,4 +205,10 @@
...
+
+  void setFlinkConfigDir(@Nullable String value);
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: As discussed in the JIRA, the better place for this is the JobServer configuration. Sorry for the hassle.

Comment (angoenka) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Sounds reasonable. I will move it their.

**Comment by tweise** [comment]
Something that may make sense as part of pipeline options is to let the user provide key value pairs of configuration options and merge those into the configuration object.

**Comment by angoenka** [comment]
Run Python Flink ValidatesRunner

**Comment by angoenka** [comment]
Run Flink ValidatesRunner

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java [hunk 3]
```diff
@@ -118,9 +119,13 @@
...
   public static BatchTranslationContext createTranslationContext(
-      JobInfo jobInfo, FlinkPipelineOptions pipelineOptions, List<String> filesToStage) {
+      JobInfo jobInfo,
+      FlinkPipelineOptions pipelineOptions,
+      @Nullable String confDir,
+      List<String> filesToStage) {
     ExecutionEnvironment executionEnvironment =
-        FlinkExecutionEnvironments.createBatchExecutionEnvironment(pipelineOptions, filesToStage);
+        FlinkExecutionEnvironments.createBatchExecutionEnvironment(
+            pipelineOptions, filesToStage, confDir);
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java]: Is the config also passed for the streaming translator?

**Comment by mxm** [comment]
I think this needs to be adapted to also use the config in streaming translation. Otherwise it looks fine!

**Comment by angoenka** [comment]
Thanks @mxm .
Updated the PR to pass the parameter in case of streaming as well. I think I missed it earlier.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkExecutionEnvironmentsTest.java [hunk 4]
```diff
@@ -84,15 +84,15 @@ public void shouldInferParallelismFromEnvironmentBatch() throws IOException {
...
+    String confDir = extractFlinkConfig();
 
     FlinkPipelineOptions options = PipelineOptionsFactory.as(FlinkPipelineOptions.class);
     options.setRunner(TestFlinkRunner.class);
     options.setFlinkMaster("host:80");
 
     StreamExecutionEnvironment sev =
         FlinkExecutionEnvironments.createStreamExecutionEnvironment(
-            options, Collections.emptyList(), flinkConfDir);
+            options, Collections.emptyList(), confDir);
```
Comment (mxm) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkExecutionEnvironmentsTest.java]: unnecessary change

**Comment by tweise** [comment]
Is this change now equivalent to setting `FLINK_CONF_DIR`? If so, it would be good to document that.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobServerDriver.java [hunk 5]
```diff
@@ -94,6 +95,18 @@ String getFlinkMasterUrl() {
     Long getSdkWorkerParallelism() {
       return this.sdkWorkerParallelism;
     }
+
+    @Option(
+      name = "--flink-conf-dir",
+      usage =
+          "Directory containing Flink YAML configuration files. These properties will be set to all jobs submitted to Flink."
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobServerDriver.java]: Mention here that it overrides `FLINK_CONF_DIR` and also confirm that if this option isn't used `FLINK_CONF_DIR` remains in effect.

Comment (angoenka) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobServerDriver.java]: Updated
